src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -207,92 +357,48 @@ std::string CHLSLCompiler::preprocessShader(std::string&& code, IShader::E_SHADE
+        logger.log("code is nullptr", system::ILogger::ELL_ERROR);
         return nullptr;
     }
     std::vector<std::string> dxc_compile_flags = {};
-    auto stage = hlslOptions.stage;
-    auto newCode = preprocessShader(code, stage, dxc_compile_flags, hlslOptions.preprocessorOptions);
+    IShader::E_SHADER_STAGE stage = options.stage;
+    auto newCode = preprocessShader(code, stage, hlslOptions.preprocessorOptions, dxc_compile_flags);
 
     // Suffix is the shader model version
-    // TODO: Figure out a way to get the shader model version automatically
-    // 
-    // We can't get it from the DXC library itself, as the different versions and the parsing
-    // use a weird lexer based system that resolves to a hash, and all of that is in a scoped variable
-    // (lib/DXIL/DxilShaderModel.cpp:83)
-    // 
-    // Another option is trying to fetch it from the commandline tool, either from parsing the help message
-    // or from brute forcing every -T option until one isn't accepted
-    //
     std::wstring targetProfile(SHADER_MODEL_PROFILE);
+   
+    std::vector<std::wstring> arguments = {};
+    if (dxc_compile_flags.size()) { // #pragma dxc_compile_flags takes priority
+        populate_arguments_with_type_conversion(arguments, dxc_compile_flags, logger);
+    }
+    else if (hlslOptions.dxcOptions.size()) { // second in order of priority is command line arguments
+        populate_arguments_with_type_conversion(arguments, hlslOptions.dxcOptions, logger);
+    }
+    else { // lastly default arguments
 
-    // Set profile two letter prefix based on stage
-    switch (stage)
-    {
-        case asset::IShader::ESS_VERTEX:
-            targetProfile.replace(0, 2, L"vs");
-            break;
-        case asset::IShader::ESS_TESSELLATION_CONTROL:
-            targetProfile.replace(0, 2, L"ds");
-            break;
-        case asset::IShader::ESS_TESSELLATION_EVALUATION:
-            targetProfile.replace(0, 2, L"hs");
-            break;
-        case asset::IShader::ESS_GEOMETRY:
-            targetProfile.replace(0, 2, L"gs");
-            break;
-        case asset::IShader::ESS_FRAGMENT:
-            targetProfile.replace(0, 2, L"ps");
-            break;
-        case asset::IShader::ESS_COMPUTE:
-            targetProfile.replace(0, 2, L"cs");
-            break;
-        case asset::IShader::ESS_TASK:
-            targetProfile.replace(0, 2, L"as");
-            break;
-        case asset::IShader::ESS_MESH:
-            targetProfile.replace(0, 2, L"ms");
-            break;
-        default:
-            hlslOptions.preprocessorOptions.logger.log("invalid shader stage %i", system::ILogger::ELL_ERROR, stage);
+        // Set profile two letter prefix based on stage
+        auto stageStr = ShaderStageToString(stage);
+        if (!stageStr) {
+            logger.log("invalid shader stage %i", system::ILogger::ELL_ERROR, stage);
             return nullptr;
```

Comment:
why have you got some duplicate logic, you're calling `try_upgrade_shader_stage` after this anyway!


Just don't include the `-T` parameter and test the fallbacks being correct in `try_upgrade_shader_stage` and `try_upgrade_hlsl_version`